### PR TITLE
go.mod: bump Pebble to 8a097e8ae86b

### PR DESCRIPTION
**Pull request by aadityasondhi**
Changes:

 * [`8a097e8a`](https://github.com/cockroachdb/pebble/commit/8a097e8a) ingest,compaction: use excise when flushing flushableIngest
 * [`dc7ccb2b`](https://github.com/cockroachdb/pebble/commit/dc7ccb2b) vfs/vfstest: adjust WithOpenFileTracking to not wrap typed nils
 * [`c268820f`](https://github.com/cockroachdb/pebble/commit/c268820f) meta: temporarily disable downloadOp
 * [`2aa3786c`](https://github.com/cockroachdb/pebble/commit/2aa3786c) sstable: split block.go
 * [`702f8cc3`](https://github.com/cockroachdb/pebble/commit/702f8cc3) base: add UserKeyBounds

Note: This commit also includes changes to the `IngestAndExcise` interface from [`8a097e8a`](https://github.com/cockroachdb/pebble/commit/8a097e8a)

Release note: none.
Epic: none.

### CODE REVIEW

**Review by RaduBerinde — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120843)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi and @itsbilal)

___
*[`pkg/kv/kvserver/replica_raftstorage.go` line 676 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/120843#-NtWwLEuBy1bvsWsa1ft:-NtWwLEuBy1bvsWsa1fu:b-bfulv7) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f97c27fe0f3e91800e7f143bc31ce623c623613/pkg/kv/kvserver/replica_raftstorage.go#L676)):*
> ```Go
> 	if inSnap.doExcise {
> 		exciseSpan := desc.KeySpan().AsRawSpanWithNoLocals()
> 		// Since these snapshot SSTs contain RANGEDELs, we can make use of storage
> ```

Do they *always* contain RANGEDELs?
___
*[`pkg/kv/kvserver/replica_raftstorage.go` line 677 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/120843#-NtWwQEPCKdaWQLhvvmN:-NtWwQEPCKdaWQLhvvmO:bagse4m) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f97c27fe0f3e91800e7f143bc31ce623c623613/pkg/kv/kvserver/replica_raftstorage.go#L677)):*
> ```Go
> 		exciseSpan := desc.KeySpan().AsRawSpanWithNoLocals()
> 		// Since these snapshot SSTs contain RANGEDELs, we can make use of storage
> 		// engine optimizations for SSTs that have the RANGEDELs.
> ```

[nit] that contain RANGEDELs
___
*[`pkg/storage/engine.go` line 1038 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/120843#-NtWw3_W3GvCsXpUs8cM:-NtWw3_W3GvCsXpUs8cN:bskih3h) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f97c27fe0f3e91800e7f143bc31ce623c623613/pkg/storage/engine.go#L1038)):*
> ```Go
> 	// IngestAndExciseFiles is a variant of IngestLocalFilesWithStats that excises
> 	// an ExciseSpan, and ingests either local or shared sstables or both. It also
> 	// takes the flag sstsContainExciseTombstone to signal that the exciseSpan
> ```

Should clarify if this flag has correctness implications - what happens if it doesn't actually contain RANGEDELs? If just an optimization, could say `that the exciseSpan contains (or probably contains) RANGEDELs`
___
*[`pkg/storage/engine.go` line 1040 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/120843#-NtWvrjx9yWiY4fpnFW1:-NtWvrjx9yWiY4fpnFW2:b-mls9qb) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f97c27fe0f3e91800e7f143bc31ce623c623613/pkg/storage/engine.go#L1040)):*
> ```Go
> 	// takes the flag sstsContainExciseTombstone to signal that the exciseSpan
> 	// contains RANGEDELs.
> 	IngestAndExciseFiles(
> ```

[nit]
```
IngestAndExciseFiles(
  ctx context.Context,
  paths []string,
  shared []pebble.SharedSSTMeta,
  external []pebble.ExternalFile,
  exciseSpan roachpb.Span,
  sstsContainExciseTombstone bool,
) (pebble.IngestOperationStats, error)
```

File: pkg/kv/kvserver/replica_raftstorage.go [hunk 1]
```diff
@@ -673,8 +673,11 @@ func (r *Replica) applySnapshot(
 	// https://github.com/cockroachdb/cockroach/issues/93251
 	if inSnap.doExcise {
 		exciseSpan := desc.KeySpan().AsRawSpanWithNoLocals()
+		// Since these snapshot SSTs contain RANGEDELs, we can make use of storage
+		// engine optimizations for SSTs that have the RANGEDELs.
+		sstsContainExciseTombstone := true
```
Comment (itsbilal) [inline pkg/kv/kvserver/replica_raftstorage.go:678]: We actually need a few other changes to make this work. Currently the SSTs don't contain rangedels. (Also specify that it's not just RangeDel but also RangeKeyDels)

1) We'll need to add a new field to inSnap that specifies if this is a `SharedOrExternalReplicate`. (We could do len(inSnap.sharedSSTs) > 0 but that is slightly wrong). `doExcise` is true for shared snapshots, external snapshots, and local snapshots. We'll be setting this to true here for local snapshots but not for shared/external snapshots.
2) See this. The last bool here specifies if the sstable should skip adding tombstones. We should set that to true for shared and external snapshots but we should flip it to false for local snapshots or the range, so we have the tombstones in the SST:

https://github.com/cockroachdb/cockroach/blob/0b2349225e63c620f66ad57226205709a058106e/pkg/kv/kvserver/store_snapshot.go#L516

**Review by itsbilal — CHANGES_REQUESTED** [review]

**Review by aadityasondhi — COMMENTED** [review]
Thanks @itsbilal and @RaduBerinde for the reviews. I addressed the comments, please take another look!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120843#-:-NtalRNb1w69w_898OqV:b-kssrew)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @itsbilal and @RaduBerinde)

___
*[`pkg/kv/kvserver/replica_raftstorage.go` line 676 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/120843#-NtWwLEuBy1bvsWsa1ft:-NtaktwnDNyXUeZXBG6T:b-n7fj1e) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f97c27fe0f3e91800e7f143bc31ce623c623613/pkg/kv/kvserver/replica_raftstorage.go#L676)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Do they *always* contain RANGEDELs?
</blockquote></details>

See the conversation with @itsbilal below. I missed the part where I did not get the RANGEDELs for free and had to make sure they exist. Made the changes to make sure they are added for the snapshots we care about and for the once that don't we set this field to false.
___
*[`pkg/kv/kvserver/replica_raftstorage.go` line 677 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/120843#-NtWwQEPCKdaWQLhvvmN:-NtalAxoEtnYytwyP-WE:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f97c27fe0f3e91800e7f143bc31ce623c623613/pkg/kv/kvserver/replica_raftstorage.go#L677)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] that contain RANGEDELs
</blockquote></details>

Done
___
*[`pkg/storage/engine.go` line 1038 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/120843#-NtWw3_W3GvCsXpUs8cM:-NtalEGy0i-hIFqDoe_E:b-306uqa) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f97c27fe0f3e91800e7f143bc31ce623c623613/pkg/storage/engine.go#L1038)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Should clarify if this flag has correctness implications - what happens if it doesn't actually contain RANGEDELs? If just an optimization, could say `that the exciseSpan contains (or probably contains) RANGEDELs`
</blockquote></details>

Thanks for pointing this out. It is the caller's responsibility to make sure that the field is appropriately assigned. Added the comment. 
___
*[`pkg/storage/engine.go` line 1040 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/120843#-NtWvrjx9yWiY4fpnFW1:-NtalQ0u28DjX7T4Y7sL:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f97c27fe0f3e91800e7f143bc31ce623c623613/pkg/storage/engine.go#L1040)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit]
```
IngestAndExciseFiles(
  ctx context.Context,
  paths []string,
  shared []pebble.SharedSSTMeta,
  external []pebble.ExternalFile,
  exciseSpan roachpb.Span,
  sstsContainExciseTombstone bool,
) (pebble.IngestOperationStats, error)
```
</blockquote></details>

Done

**Comment by aadityasondhi** [comment]
Same error in CI as https://github.com/cockroachdb/pebble/pull/3437#issuecomment-2013489899
```
MANIFEST apply failed: pebble: internal error: L6 files 000009 and 000012 have overlapping ranges: [/Local/RangeID/34/u""/0,0#21,RANGEDEL-/Local/RangeID/34/v""/0,0#inf,RANGEDEL] vs [/Local/RangeID/34/u/RaftHardState/0,0#19,SET-/Local/Store"uptm"/0,0#14,SET]
```

**Review by jbowens — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120843)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi, @itsbilal, and @RaduBerinde)

___
*[`pkg/storage/pebble.go` line 2147 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120843#-NtbPDMDDBR_CoESDZDU:-NtbPDME0aHsjxOTlTXc:b-5y6hyl) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7f778d57ea2b7b4e815da4ca53901598ca071a5/pkg/storage/pebble.go#L2147)):*
> ```Go
> 		End:   EngineKey{Key: exciseSpan.EndKey}.Encode(),
> 	}
> 	return p.db.IngestAndExcise(paths, shared, external, rawSpan, sstsContainExciseTombstone)
> ```

would we be able to just pass `false` here to unblock the bump and begin using this new functionality in a follow-up PR?

**Review by aadityasondhi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120843)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @itsbilal, @jbowens, and @RaduBerinde)

___
*[`pkg/storage/pebble.go` line 2147 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120843#-NtbPDMDDBR_CoESDZDU:-NtbPaUO2iKtWRAHVTlg:b-b6o137) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7f778d57ea2b7b4e815da4ca53901598ca071a5/pkg/storage/pebble.go#L2147)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

would we be able to just pass `false` here to unblock the bump and begin using this new functionality in a follow-up PR?
</blockquote></details>

Yes, sounds reasonable. I will do that now.

**Comment by aadityasondhi** [comment]
@jbowens @itsbilal please take a look at this now. I have disabled the new changes and have added TODOs wherever we need to add code back once the bugs are sorted.

**Review by RaduBerinde — APPROVED** [review]

**Pull request closed without merging**